implementations/micrometer-registry-graphite/src/main/java/io/micrometer/graphite/GraphiteTagNameMapper.java

```diff
@@ -0,0 +1,61 @@
+/**
+ * Copyright 2017 Pivotal Software, Inc.
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * https://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micrometer.graphite;
+
+import io.micrometer.core.instrument.Meter;
+import io.micrometer.core.instrument.Tag;
+import io.micrometer.core.instrument.config.NamingConvention;
+import io.micrometer.core.instrument.util.HierarchicalNameMapper;
+
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * Tag based {@link HierarchicalNameMapper} for Graphite.
+ *
+ * @author Jon Schneider
+ * @author Johnny Lim
+ * @author Andrew Fitzgerald
+ */
+public class GraphiteTagNameMapper implements HierarchicalNameMapper {
```

Comment:
Once we have tags, is it really a hierarchical metric?

If we say it is, that makes it easier to reuse the existing graphite registry and just sub in the new name mapper....

If it doesn't there's a lot more work to do since we're also likely no longer using the dropwizard registry as our base (right?)

Reply:
First impression: If this tag name mapper is the default `HierarchicalNameMapper` (which I think makes sense), then most people would never even interact with the type because they'd use one of the other constructors like `GraphiteMeterRegistry(GraphiteConfig config, Clock clock)`.

And I think it makes sense for this to be a specialization of `HierarchicalNameMapper`, because unlike purely dimensional monitoring systems, Graphite doesn't _require_ the use of this tags support. Even folks using Graphite >1.1 may have their reasons for avoiding it (e.g. consistency with existing names/dashboards?).

So I think you have done the right thing.